src/Framework/Core/Controls/Options/PagingImplementation.cs

```diff
@@ -17,6 +23,74 @@ public static IQueryable<T> ApplyPagingToQueryable<T, TPagingOptions>(IQueryable
                 ? queryable.Skip(options.PageSize * options.PageIndex).Take(options.PageSize)
                 : queryable;
         }
-    }
 
+#if NET6_0_OR_GREATER
+        /// <summary> Attempts to count the queryable asynchronously. EF Core IQueryables are supported, and IQueryables which return IAsyncEnumerable from GroupBy operator also work correctly. Otherwise, a synchronous fallback is used, or <see cref="CustomAsyncQueryableCountDelegate" /> may be set to add support for an ORM mapper of choice. </summary>
+        public static async Task<int> QueryableAsyncCount<T>(IQueryable<T> queryable, CancellationToken ct = default)
+        {
+            if (CustomAsyncQueryableCountDelegate is {} customDelegate)
+            {
+                var result = await customDelegate(queryable, ct);
+                if (result.HasValue)
+                {
+                    return result.Value;
+                }
+            }
+
+            var queryableType = queryable.GetType();
+            // Note: there is not a standard way to get a count from IAsyncEnumerable instance, without enumerating it.
+            //       we use two heuristics to try to get the count using the query provider:
+            //          * if we detect usage of EF Core, call its CountAsync method
+            //          * otherwise, do it as .GroupBy(_ => 1).Select(group => group.Count()).SingleOrDefault()
+            //  (if you are reading this and need a separate hack for your favorite ORM, you can set
+            //   CustomAsyncQueryableCountDelegate, and we do accept PRs adding new heuristics ;) )
+            return await (
+                EfCoreAsyncCountHack(queryable, queryableType, ct) ??
```

Comment:
It would be great to add Marten.

Reply:
Yes, I have created an issue for that: #1927. It's slightly non-trivial to test this, as Marten needs a running postgres